src/foam/nanos/crunch/ServerCrunchService.java

```diff
@@ -205,6 +207,51 @@ public UserCapabilityJunction getJunction(X x, String capabilityId) {
     return this.getJunctionForSubject(x, capabilityId, subject);
   }
 
+  public boolean hasPreconditionsMet(
+    X x, String capabilityId
+  ) {
+    // Return false if capability does not exist or is not available
+    var capabilityDAO = (DAO) x.get("capabilityDAO");
+    if ( capabilityDAO.inX(x).find(capabilityId) == null ) return false;
+
+    var preconditions = Arrays.stream(((CapabilityCapabilityJunction[]) ((ArraySink) ((DAO) x.get("prerequisiteCapabilityJunctionDAO"))
+      .where(AND(
+        EQ(CapabilityCapabilityJunction.SOURCE_ID, capabilityId),
+        EQ(CapabilityCapabilityJunction.PRECONDITION, true)
+      ))
+      .select(new ArraySink())).getArray()
+      .toArray(new CapabilityCapabilityJunction[0])))
+      .map(CapabilityCapabilityJunction::getTargetId).toArray(String[]::new);
```

Comment:
please use MAP sink instead.

Reply:
I created a ticket for this: https://nanopay.atlassian.net/browse/NP-2907

I think it might make sense to optimize this method later so I can consider caching possibilities at the same time.

Reply:
I agree with @lchanmann.

Reply:
@KernelDeimos can you add a TODO comment if we are going to defer this?